[browser][websockets] Support for CloseOutputAsync

kjpou1 opened the pull request:
Note this relies on PR : https://github.com/dotnet/runtime/pull/45827

- Add support for CloseOutputAsync
- Remove `[ActiveIssue("https://github.com/dotnet/runtime/issues/45468", TestPlatforms.Browser)]`
- Add more tests

Fix https://github.com/dotnet/runtime/issues/46909
Fix https://github.com/dotnet/runtime/issues/45468
Fix https://github.com/dotnet/runtime/issues/45674
Fix https://github.com/dotnet/runtime/issues/45586
Close https://github.com/dotnet/runtime/issues/46983

Comments from parent PR:

Multiple fixes for Abort.

There are multiple failures with the Abort tests causing some tests to:

- Never end
- Multiple GC calls that end in running out of memory. 

- Some failures are expected to fail with specific WebSocketError codes:

Error codes that are not correct tend to be when websockets tests are looking for specific error codes and messages when the browser returns `NativeError` and the websocket native error code instead.

```
failed: WebSocket is closed before the connection is established.
```

---

These need to be marked correctly and or caught and wrapped correctly.

```
Assert.Equal() Failure
  info: Expected: Aborted
  info: Actual:   Closed
```

Fixes for these should also fix some, note not all, of the memory runtime errors as per issue: https://github.com/dotnet/runtime/issues/45586

--- 

Fix tests that were failing when expecting CloseSent as a valid state.  

```
// fail: [FAIL] System.Net.WebSockets.Client.Tests(server: ws://corefx-net-http11.azurewebsites.net/WebSocket/EchoWebSocket.ashx)
//   info: Assert.Throws() Failure
//   info: Expected: typeof(System.OperationCanceledException)
//   info: Actual:   typeof(System.Net.WebSockets.WebSocketException): The WebSocket is in an invalid state ('Aborted') for this operation. Valid states are: 'Open, CloseSent'
```

---

Tests are intermittently failing with System.OperationCanceledException : The operation was canceled.

~~Added ActiveIssue for these two tests.  https://github.com/dotnet/runtime/issues/46909~~

```
It seems that sometimes the default timeout is not enough for browser so it is extended for browser environments
See issue https://github.com/dotnet/runtime/issues/46909
```

Further addressed in https://github.com/dotnet/runtime/pull/47906


---

SendReceive_PartialMessageBeforeCompleteMessageArrives_Success which describes the test as Ask the remote server to echo back received messages without ever signaling "end of message".  This test hangs as javascript fetch does not support this functionality.

See issue https://github.com/dotnet/runtime/issues/46983 for more detailed description and native javascript tests.

---

Associate PR's: https://github.com/dotnet/runtime/issues/44666 and https://github.com/dotnet/runtime/pull/45470

Close https://github.com/dotnet/runtime/issues/45586
Fix https://github.com/dotnet/runtime/issues/45674

kjpou1 committed:
Set the WebSocketException to Faulted to fix tests

kjpou1 committed:
Fix the exceptions that are being thrown to coincide with test expectations

kjpou1 committed:
Fix Dispose we are not processing it multiple times.

kjpou1 committed:
Fix Abort code so the messages align correctly with the tests.

example from tests:

```
Assert.Equal() Failure
  info: Expected: Aborted
  info: Actual:   Closed
```

kjpou1 committed:
Set the WebSocketException to Faulted to fix test expectations

kjpou1 committed:
Close the connections correctly.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Fix Abort test Abort_CloseAndAbort_Success

- This was causing a never ending Task when aborted after a Close already executed.
- Never ended which was a cause of memory errors after left running.
- See also https://github.com/dotnet/runtime/issues/45586

kjpou1 committed:
- Fixes for ReceiveAsyncTest

- Exception text not sent correctly.  This test was expecting 'Aborted'.
- Mismatched expected exception messages
- Make sure the connection is torn down.
- Multiple GC calls that end in running out of memory fixed.  https://github.com/dotnet/runtime/issues/45586

```
//  [FAIL] System.Net.WebSockets.Client.Tests.CancelTest.ReceiveAsync_AfterCancellationDoReceiveAsync_ThrowsWebSocketException(server: ws://corefx-net-http11.azurewebsites.net/WebSocket/EchoWebSocket.ashx)
//   info: Assert.Equal() Failure
//   info:                                  ↓ (pos 39)
//   info: Expected: ··· an invalid state ('Aborted') for this operation. Valid state···
//   info: Actual:   ··· an invalid state ('Open') for this operation. Valid states a···
//   info:                                  ↑ (pos 39)

```

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

# Conflicts:
#	src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

lewing committed:
Remove ActiveIssue attributes that should be fixed

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Add back code after merge conflict

- Update tests that are resolved.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Fix tests that were failing when expecting CloseSent as a valid state.

```
// fail: [FAIL] System.Net.WebSockets.Client.Tests(server: ws://corefx-net-http11.azurewebsites.net/WebSocket/EchoWebSocket.ashx)
//   info: Assert.Throws() Failure
//   info: Expected: typeof(System.OperationCanceledException)
//   info: Actual:   typeof(System.Net.WebSockets.WebSocketException): The WebSocket is in an invalid state ('Aborted') for this operation. Valid states are: 'Open, CloseSent'
```

kjpou1 committed:
Fix the Abort and Cancel never ending tests.

- Fix the clean up of the task and set or cancel the task cleanly.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Add ActiveIssue to websocket client SendRecieve tests

kjpou1 committed:
Add ActiveIssue to websocket client SendRecieve tests

- intermittently failing with System.OperationCanceledException : The operation was canceled.

kjpou1 committed:
Add extra time to timeout for a couple of tests that were intermittently failing with System.OperationCanceledException : The operation was canceled.

- This was an ActiveIssue https://github.com/dotnet/runtime/issues/46909 but extending the time seems to do the job.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Add ActiveIssue to websocket client SendRecieve tests

- [browser][websocket] Hang with responses without ever signaling "end of message"
- [ActiveIssue("https://github.com/dotnet/runtime/issues/46983", TestPlatforms.Browser)]  // JS Fetch does not support see issue

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Remove `Interlocked` code as per review comment.

kjpou1 committed:
Fix comment

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Fix Abort tests on non chrome browsers.

- Fix for those browsers that do not set Close and send an onClose event in certain instances i.e. firefox and safari.
- Chrome will send an onClose event and we tear down the websocket there.  Other browsers need to be handled differently.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-abort-fixes

kjpou1 committed:
Merge branch 'master' of https://github.com/kjpou1/runtime into wasm-browser-websocket-CloseOutputAsync

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-CloseOutputAsync

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-CloseOutputAsync

kjpou1 committed:
We should not throw here.

- Closing an already-closed ClientWebSocket should be a no-op.

kjpou1 committed:
Add `CloseOutputAsync` implementation

- The browser websocket implementation does not support this so we fake it the best we can.

kjpou1 committed:
Remove `ActiveIssue` and add more tests

- Remove `[ActiveIssue("https://github.com/dotnet/runtime/issues/45468", TestPlatforms.Browser)]`
- Add more tests

lewing commented:
/azp run runtime-libraries-mono outerloop

BrennanConroy commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -498,7 +575,38 @@ private Task CloseAsyncCore(WebSocketCloseStatus closeStatus, string? statusDesc
+            {
+                _innerWebSocketCloseStatus = closeStatus;
+                _innerWebSocketCloseStatusDescription = statusDescription;
+                _innerWebSocket!.Invoke("close", (int)closeStatus, statusDescription);
```
I'd just like to reiterate that not calling `onClose` manually here could result in a large delay. https://github.com/dotnet/runtime/issues/45468#issuecomment-741519725

kjpou1 replied:
Brennan I saw that but is there anyway we can test this easily?

Here is the sequence of event on CloseOutputAsync that is proposed.

- Remove all event delegates not just on the close event or is close event enough?
- Call JavaScript `close()` method on `websocket`.
- Call JavaScript `onClose` event manually on the `websocket`.

I ask because if I leave the code as is the `onClose` will be called as well but the dotnet method returns immediately.

BrennanConroy replied:
We clear all events on the `websocket` (including `onClose`), call `websocket.close()` then call the user provided `onClose` method manually.

If you can write a test that blocks in `websocket.close()` and checks that the `onClose` method is still called then that would be good.

kjpou1 replied:
Have to admit not sure how that would be done from the test harness for Websocket Client.  

Will add the code above.  Thanks for the feedback.

kjpou1 replied:
Addressed if recent commits

kjpou1 commented:
The errors in the outerloop are for the following:

```
error : Work item System.Memory.Tests
error : Work item System.Text.Json.Tests
error : Work item System.Threading.Tasks.Parallel.Tests
error : Work item System.Threading.Tasks.Tests 
```

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-CloseOutputAsync

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-CloseOutputAsync

wfurt commented on src/libraries/System.Net.WebSockets.Client/tests/ClientWebSocketTestBase.cs:
```diff
@@ -26,6 +26,7 @@ public class ClientWebSocketTestBase
         }).ToArray();
 
         public const int TimeOutMilliseconds = 20000;
+        public const int BrowserTimeOutMilliseconds = 30000;
```
as far as I can tell, this is just upper boundary. I think it would be ok to just bump up TimeOutMilliseconds instead of creating new variable. If you feel this should be different, I would put the logic here instead of dragging Browser branches through the tests.

kjpou1 replied:
Addressed in recent commits

wfurt commented on src/libraries/System.Net.WebSockets.Client/tests/CloseTest.cs:
```diff
@@ -173,15 +215,16 @@ public async Task CloseOutputAsync_ClientInitiated_CanReceive_CanClose(Uri serve
                 // available before calling close. The WinRT MessageWebSocket implementation doesn't allow receiving
                 // after a call to Close.
-                await Task.Delay(100);
+                await Task.Delay(PlatformDetection.IsBrowser ? 500 : 100); // default timeout is not enough for browser sometimes so we will extend it
```
could we just use 500 for all platforms? This is already OuterLoop test so it may not matter much.

kjpou1 replied:
Addressed in recent commits

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-browser-websocket-CloseOutputAsync

kjpou1 committed:
Address timeout implementation review

kjpou1 committed:
Add code as per SignalR comments to prevent long wait times in some cases

As per comments
- We clear all events on the websocket (including onClose),
- call websocket.close()
- then call the user provided onClose method manually.

kjpou1 commented:
/azp run runtime-libraries-mono outerloop

wfurt reviewed (approved):
LGTM

lewing commented on src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs:
```diff
@@ -88,7 +88,7 @@ public async Task SendReceive_PartialMessageDueToSmallReceiveBuffer_Success(Uri
         [OuterLoop("Uses external servers")]
         [ConditionalTheory(nameof(WebSocketsSupported)), MemberData(nameof(EchoServers))]
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/45586", TestPlatforms.Browser)]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/46983", TestPlatforms.Browser)]  // JS Websocket does not support see issue
```
Do we intend to fix this?  If not it should be a platform check and not an ActiveIssue

kjpou1 replied:
We can not fix this as it is.  In the issue with the JavaScript code it does not work either.  If that does not work and the network layer does not have a response there is no way to fix it in the managed code.

kjpou1 replied:
Also left it as an active issue to help document why.

kjpou1 replied:
Addressed in recent commits

lewing commented on src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs:
```diff
@@ -329,6 +326,7 @@ public async Task SendReceive_VaryingLengthBuffers_Success(Uri server)
             using (ClientWebSocket cws = await WebSocketHelper.GetConnectedWebSocket(server, TimeOutMilliseconds, _output))
             {
                 var rand = new Random();
+
```
```suggestion
```

lewing reviewed (approved):
Looks good, mentioned a couple of nits.

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs

Co-authored-by: Larry Ewing <lewing@microsoft.com>

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -112,19 +120,24 @@ readyState switch
-            else if (priorState != (int)InternalState.Created)
+            // Check that we have not started already.
+            int prevState = _state;
+            _state = _state == (int)InternalState.Created ? (int)InternalState.Connecting : _state;
```
```suggestion
            if (prevState == (int)InternalState.Created)
            {
                _state = (int)InternalState.Connecting;
            }
```
?

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -236,6 +227,37 @@ internal async Task ConnectAsyncJavaScript(Uri uri, CancellationToken cancellati
         }
 
+
+        private void onCloseCallback(JSObject? closeEvt, CancellationToken cancellationToken)
```
Nit: Can we use PascalCasing for members like this?

kjpou1 replied:
Addressed in recent commits

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -192,19 +181,21 @@ internal async Task ConnectAsyncJavaScript(Uri uri, CancellationToken cancellati
                         else
                         {
-                            tcsConnect.SetCanceled(cancellationToken);
+                            _tcsConnect.SetCanceled(cancellationToken);
```
Elsewhere you've used TrySet*... is it intentional for these two that they're using the non-Try variants?

kjpou1 replied:
Addressed in recent commits

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -318,32 +340,51 @@ private void NativeCleanup()
+                if (!_cts.IsCancellationRequested)
+                {
+                    // registered by the CancellationTokenSource cts in the connect method
+                    _cts.Cancel(false);
```
```suggestion
                    _cts.Cancel();
```

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

kjpou1 committed:
Address review comments for using `TrySet*` variants

kjpou1 committed:
Address review about using PascalCasing

kjpou1 committed:
Change test to be a platform check and not an ActiveIssue as per review
